File: build_swpackage/hex2bin.py
#!/usr/bin/env python3
import sys
import argparse
from intelhex import IntelHex

def main():
    parser = argparse.ArgumentParser(
        description='Convert an Intel HEX file to a binary file, filling gaps with 0.'
    )
    parser.add_argument('input_hex', help='Path to input Intel HEX file')
    parser.add_argument('start_address', help='Start address (e.g., 0x80100000)')
    parser.add_argument('end_address', help='End address (e.g., 0x80200000)')
    parser.add_argument('output_bin', help='Path to output binary file')

    args = parser.parse_args()

    # Convert start and end addresses to integers. The base 0 allows hex values (e.g., 0x80100000)
    try:
        start_addr = int(args.start_address, 0)
        end_addr = int(args.end_address, 0)
    except ValueError as e:
        print("Error: Could not convert addresses to numbers.", e)
        sys.exit(1)

    if start_addr > end_addr:
        print("Error: Start address must be less than or equal to end address.")
        sys.exit(1)

    # Create an instance of IntelHex and load the HEX file.
    ih = IntelHex()
    try:
        ih.loadhex(args.input_hex)
    except Exception as e:
        print("Error: Could not load HEX file:", e)
        sys.exit(1)

    # Write the binary file using the 'start' and 'end' keywords and pad with 0.
    try:
        ih.tobinfile(args.output_bin, start=start_addr, end=end_addr, pad=0)
        print(f"Successfully wrote binary file: {args.output_bin}")
    except Exception as e:
        print("Error: Could not write binary file:", e)
        sys.exit(1)

if __name__ == "__main__":
    main()

File: archive/hex2bin.js
#!/usr/bin/env node

const fs = require('fs');
const intelHex = require('intel-hex');

function parseAddress(addrStr) {
  const addr = Number(addrStr);
  if (isNaN(addr)) {
    if (addrStr.startsWith('0x') || addrStr.startsWith('0X')) {
      return parseInt(addrStr, 16);
    }
    throw new Error(`Invalid address: ${addrStr}`);
  }
  return addr;
}

function usage() {
  console.log('Usage: node hex2bin.js <input_hex> <start_address> <end_address> <output_bin>');
  process.exit(1);
}

async function main() {
  const args = process.argv.slice(2);
  if (args.length !== 4) {
    usage();
  }

  const [inputHexPath, startAddressStr, endAddressStr, outputBinPath] = args;

  try {
    const startAddr = parseAddress(startAddressStr);
    const endAddr = parseAddress(endAddressStr);

    if (startAddr > endAddr) {
      console.error('Error: Start address must be less than or equal to end address.');
      process.exit(1);
    }

    const hexContent = fs.readFileSync(inputHexPath, 'utf8');
    const parsed = intelHex.parse(hexContent);

    const baseAddr = parsed.startAddress || 0;
    const length = endAddr - startAddr;
    if (length <= 0) {
      console.error('Error: End address must be greater than start address.');
      process.exit(1);
    }

    const outputBuffer = Buffer.alloc(length, 0);

    const dataStart = baseAddr;
    const dataEnd = baseAddr + parsed.data.length;

    const copyStart = Math.max(startAddr, dataStart);
    const copyEnd = Math.min(endAddr, dataEnd);

    if (copyEnd > copyStart) {
      const sourceStartIndex = copyStart - dataStart;
      const destStartIndex = copyStart - startAddr;
      const bytesToCopy = copyEnd - copyStart;

      parsed.data.copy(outputBuffer, destStartIndex, sourceStartIndex, sourceStartIndex + bytesToCopy);
    }

    fs.writeFileSync(outputBinPath, outputBuffer);
    console.log(`Successfully wrote binary file: ${outputBinPath}`);

  } catch (err) {
    console.error('Error:', err.message || err);
    process.exit(1);
  }
}

main();
